flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetAggregate.scala

```diff
@@ -69,37 +72,55 @@ class DataSetGroupReduce(
 
     expectedType match {
       case Some(typeInfo) if typeInfo.getTypeClass != classOf[Row] =>
-        throw new PlanGenException("GroupReduce operations currently only support returning Rows.")
+        throw new PlanGenException("Aggregate operations currently only support returning Rows.")
       case _ => // ok
     }
 
+    val groupingKeys = (0 until grouping.length).toArray
+    // add grouping fields, position keys in the input, and input type
+    val aggregateResult = AggregateUtil.createOperatorFunctionsForAggregates(namedAggregates,
+      inputType, rowType, grouping)
+
     val inputDS = input.asInstanceOf[DataSetRel].translateToPlan(
       config,
       // tell the input operator that this operator currently only supports Rows as input
       Some(TypeConverter.DEFAULT_ROW_TYPE))
 
+    val intermediateType = determineReturnType(
+      aggregateResult.intermediateDataType,
+      expectedType,
+      config.getNullCheck,
+      config.getEfficientTypeUsage)
+
+
     // get the output types
-    val fieldsNames = rowType.getFieldNames
     val fieldTypes: Array[TypeInformation[_]] = rowType.getFieldList.asScala
     .map(f => f.getType.getSqlTypeName)
     .map(n => TypeConverter.sqlTypeToTypeInfo(n))
     .toArray
 
     val rowTypeInfo = new RowTypeInfo(fieldTypes)
+
+    val mappedInput = inputDS.map(aggregateResult.mapFunc.apply(
```

Comment:
We could refactor the `AggregateUtil` such that it returns the `MapFunction` directly and not a function to generate the `MapFunction`. I think would not need to expose the `intermediateDataType` of the `aggregateResult`.